app/queries/get_activity_log_events.rb

```diff
@@ -0,0 +1,20 @@
+class GetActivityLogEvents
+  def self.call(application_choices:, since: nil)
+    since ||= Time.zone.local(2018, 1, 1) # before the pilot began, i.e. all records
+
+    Audited::Audit.from <<~COMBINE_AUDITS_WITH_APPLICATION_CHOICES_SCOPE.squish
+      (
+        SELECT a.*
+          FROM audits a
+          INNER JOIN application_choices ac
+            ON auditable_id = ac.id
+              AND auditable_type = 'ApplicationChoice'
+              AND action = 'update'
+          INNER JOIN (#{application_choices.to_sql}) visible
```

Comment:
As Ben has already hinted at, the `INNER JOIN (#{application_choices.to_sql})` segment is the most important part of this query. It allows to search for `Audited::Audit` records while limiting them to the ones associated with application choices from the `GetApplicationChoicesForProviders` query.

There are multiple challenges with this query: a. `Audited::Audit` has a polymorphic association with `ApplicationChoice`, not a direct one, b. we need to somehow combine the query from `GetApplicationChoicesForProviders` into this one and c. this external query is `ApplicationChoice`-based but we need an `Audited::Audit`-based query overall.

Because the root of the query is audits, the multiplexing and ordering of events across applications is straightforward and fast. Considering the main aim of this story is speed, I'm happy to sacrifice some readability for performance.

Reply:
Call `GetApplicationChoicesForProvider.call` seems to run that query whatever — ie it's a fixed cost we'll need to take anyway. Is there a way to not run that query in advance, and if not, could we replace this `JOIN` with a `where`?

Reply:
Ah — no! I think `pry` is forcing it to realise when it prints the result. `GetApplicationChoicesForProviders.call(providers: pu.providers);` doesn't run the query.